service/resource_router_test.go

```diff
@@ -0,0 +1,90 @@
+package service
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/kvm-operator/service/resource/configmapv3"
+	"github.com/giantswarm/kvm-operator/service/resource/servicev2"
+	"github.com/giantswarm/operatorkit/framework"
+)
+
+func Test_NewResourceRouter(t *testing.T) {
+	resourceV1 := []framework.Resource{
+		&servicev2.Resource{},
+	}
+	resourceV2 := []framework.Resource{
+		&servicev2.Resource{},
+		&configmapv3.Resource{},
+	}
+
+	testCases := []struct {
+		ctx                context.Context
+		versionedResources map[string][]framework.Resource
+		customObject       interface{}
+		expectedResources  []framework.Resource
+	}{
+		// Test 1, get resource
+		{
+			ctx: context.TODO(),
+			versionedResources: map[string][]framework.Resource{
+				"1.0.0": resourceV1,
+			},
+			customObject: &v1alpha1.KVMConfig{
+				Spec: v1alpha1.KVMConfigSpec{
+					VersionBundle: v1alpha1.KVMConfigSpecVersionBundle{
+						Version: "1.0.0",
+					},
+				},
+			},
+			expectedResources: resourceV1,
+		},
+		// Test 2, get correct resources on multiples resources version
+		{
+			ctx: context.TODO(),
+			versionedResources: map[string][]framework.Resource{
+				"1.0.0": resourceV1,
+				"2.0.0": resourceV2,
+			},
+			customObject: &v1alpha1.KVMConfig{
+				Spec: v1alpha1.KVMConfigSpec{
+					VersionBundle: v1alpha1.KVMConfigSpecVersionBundle{
+						Version: "2.0.0",
+					},
+				},
+			},
+			expectedResources: resourceV2,
+		},
+		// Test 3, get resources from empty Version Bundle
+		{
+			ctx: context.TODO(),
+			versionedResources: map[string][]framework.Resource{
+				"1.0.0": resourceV1,
+				"2.0.0": resourceV2,
+				"":      resourceV1,
+			},
+			customObject: &v1alpha1.KVMConfig{
+				Spec: v1alpha1.KVMConfigSpec{
+					VersionBundle: v1alpha1.KVMConfigSpecVersionBundle{
+						Version: "",
+					},
+				},
+			},
+			expectedResources: resourceV1,
+		},
+	}
```

Comment:
The tests do not cover error cases. In the tests in the aws-operator we also have error matchers. I would expect a test case here where the router is configured with version bundles and a custom object that does not have a version bundle version. This is actually the case that failed in production, so the tests should cover this case and make sure it fails if the router is not configured properly.

Reply:
Done